File: elixir/e2e/testharness/proxy.ex
defmodule CopilotE2E.TestHarness.Proxy do
  @moduledoc """
  Replaying CAPI proxy for E2E tests.

  Manages a child process that acts as a replaying proxy to AI endpoints.
  Spawns the shared test harness server from `test/harness/server.ts`.
  """

  use GenServer

  require Logger

  # ---------------------------------------------------------------------------
  # Public API
  # ---------------------------------------------------------------------------

  @doc "Start the proxy GenServer (does not yet launch the node process)."
  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, :ok, opts)
  end

  @doc """
  Launch the replay proxy server and return its URL.

  Spawns `npx tsx test/harness/server.ts` and waits for the
  `Listening: http://...` line on stdout.
  """
  @spec start_proxy(GenServer.server()) :: {:ok, String.t()} | {:error, any()}
  def start_proxy(server) do
    GenServer.call(server, :start_proxy, 30_000)
  end

  @doc """
  Gracefully shut down the replay proxy.

  Sends a POST to `/stop` and waits for the OS process to exit.
  """
  @spec stop_proxy(GenServer.server(), keyword()) :: :ok
  def stop_proxy(server, opts \\ []) do
    GenServer.call(server, {:stop_proxy, opts}, 15_000)
  catch
    :exit, _ -> :ok
  end

  @doc "Return the proxy URL (nil if not started)."
  @spec url(GenServer.server()) :: String.t() | nil
  def url(server) do
    GenServer.call(server, :url)
  end

  @doc """
  Configure the proxy for a specific test snapshot.

  Sends a POST to `/config` with the snapshot file path and working directory.
  """
  @spec configure(GenServer.server(), String.t(), String.t()) :: :ok | {:error, any()}
  def configure(server, file_path, work_dir) do
    GenServer.call(server, {:configure, file_path, work_dir}, 10_000)
  end

  # ---------------------------------------------------------------------------
  # GenServer callbacks
  # ---------------------------------------------------------------------------

  defmodule State do
    @moduledoc false
    defstruct port: nil, proxy_url: nil, os_pid: nil
  end

  @impl true
  def init(:ok) do
    {:ok, %State{}}
  end

  @impl true
  def handle_call(:start_proxy, _from, %State{proxy_url: url} = state) when url != nil do
    {:reply, {:ok, url}, state}
  end

  def handle_call(:start_proxy, _from, %State{proxy_url: nil} = state) do
    case launch_proxy() do
      {:ok, port, proxy_url, os_pid} ->
        new_state = %State{state | port: port, proxy_url: proxy_url, os_pid: os_pid}
        {:reply, {:ok, proxy_url}, new_state}

      {:error, reason} ->
        {:reply, {:error, reason}, state}
    end
  end

  def handle_call({:stop_proxy, opts}, _from, state) do
    new_state = do_stop(state, opts)
    {:reply, :ok, new_state}
  end

  def handle_call(:url, _from, state) do
    {:reply, state.proxy_url, state}
  end

  def handle_call({:configure, file_path, work_dir}, _from, state) do
    if state.proxy_url do
      body = Jason.encode!(%{"filePath" => file_path, "workDir" => work_dir})
      url = "#{state.proxy_url}/config"

      case http_post(url, body) do
        {:ok, 200, _body} -> {:reply, :ok, state}
        {:ok, status, _body} -> {:reply, {:error, "config failed with status #{status}"}, state}
        {:error, reason} -> {:reply, {:error, reason}, state}
      end
    else
      {:reply, {:error, "proxy not started"}, state}
    end
  end

  @impl true
  def handle_info({port, {:data, _data}}, %State{port: port} = state) do
    # Drain stdout/stderr from the proxy process after startup
    {:noreply, state}
  end

  def handle_info({port, {:exit_status, _status}}, %State{port: port} = state) do
    Logger.debug("Proxy process exited")
    {:noreply, %State{state | port: nil, proxy_url: nil, os_pid: nil}}
  end

  def handle_info(_msg, state) do
    {:noreply, state}
  end

  @impl true
  def terminate(_reason, state) do
    do_stop(state, [])
    :ok
  end

  # ---------------------------------------------------------------------------
  # Private helpers
  # ---------------------------------------------------------------------------

  defp launch_proxy do
    server_path = server_ts_path()

    unless File.exists?(server_path) do
      raise "Test harness server not found at #{server_path}"
    end

    harness_dir = Path.dirname(server_path)

    # On Windows, spawn npx via cmd.exe so that npx.cmd is resolved
    {executable, args} =
      case :os.type() do
        {:win32, _} ->
          {System.find_executable("cmd") || "cmd.exe",
           ["/c", "npx", "tsx", server_path]}

        _ ->
          npx = System.find_executable("npx") || "npx"
          {npx, ["tsx", server_path]}
      end

    port_opts = [
      :binary,
      :exit_status,
      :use_stdio,
      :stderr_to_stdout,
      {:args, args},
      {:cd, String.to_charlist(harness_dir)}
    ]

    port = Port.open({:spawn_executable, String.to_charlist(executable)}, port_opts)

    case read_listening_url(port) do
      {:ok, proxy_url} ->
        # Extract OS PID from port info for cleanup
        os_pid =
          case Port.info(port, :os_pid) do
            {:os_pid, pid} -> pid
            _ -> nil
          end

        {:ok, port, proxy_url, os_pid}

      {:error, reason} ->
        safe_close_port(port)
        {:error, reason}
    end
  end

  defp read_listening_url(port) do
    read_listening_url(port, "", 15_000)
  end

  defp read_listening_url(_port, _buffer, timeout) when timeout <= 0 do
    {:error, "timed out waiting for proxy to start"}
  end

  defp read_listening_url(port, buffer, timeout) do
    start = System.monotonic_time(:millisecond)

    receive do
      {^port, {:data, data}} ->
        new_buffer = buffer <> data

        case Regex.run(~r/Listening: (http:\/\/[^\s]+)/, new_buffer) do
          [_, url] ->
            {:ok, String.trim(url)}

          nil ->
            elapsed = System.monotonic_time(:millisecond) - start
            read_listening_url(port, new_buffer, timeout - elapsed)
        end

      {^port, {:exit_status, status}} ->
        {:error, "proxy exited with status #{status} before printing URL. Output: #{buffer}"}
    after
      timeout ->
        {:error, "timed out waiting for proxy to start. Output so far: #{buffer}"}
    end
  end

  defp do_stop(%State{proxy_url: nil} = state, _opts), do: state

  defp do_stop(%State{} = state, opts) do
    skip_cache = Keyword.get(opts, :skip_writing_cache, false)

    # Send graceful stop request
    if state.proxy_url do
      stop_url =
        if skip_cache,
          do: "#{state.proxy_url}/stop?skipWritingCache=true",
          else: "#{state.proxy_url}/stop"

      try do
        http_post(stop_url, "")
      rescue
        _ -> :ok
      catch
        _, _ -> :ok
      end
    end

    # Wait briefly for the process to exit, then force-close the port
    if state.port do
      receive do
        {port, {:exit_status, _}} when port == state.port -> :ok
      after
        3_000 -> safe_close_port(state.port)
      end
    end

    %State{state | port: nil, proxy_url: nil, os_pid: nil}
  end

  defp safe_close_port(port) do
    try do
      Port.close(port)
    rescue
      _ -> :ok
    catch
      _, _ -> :ok
    end
  end

  defp server_ts_path do
    # From elixir/e2e/testharness/ go up to repo root: ../../../test/harness/server.ts
    Path.expand("../../../test/harness/server.ts", __DIR__)
  end

  # Minimal HTTP POST using Erlang's built-in :httpc client (no extra deps)
  defp http_post(url, body) do
    :inets.start()
    :ssl.start()

    headers = [
      {~c"content-type", ~c"application/json"}
    ]

    request = {String.to_charlist(url), headers, ~c"application/json", String.to_charlist(body)}

    case :httpc.request(:post, request, [{:timeout, 5_000}], []) do
      {:ok, {{_, status, _}, _headers, resp_body}} ->
        {:ok, status, List.to_string(resp_body)}

      {:error, reason} ->
        {:error, reason}
    end
  end
end

File: lua/e2e/testharness/proxy.lua
--- Replaying CAPI proxy for E2E tests.
--
-- Manages a child process that acts as a replaying proxy to AI endpoints.
-- Spawns the shared test harness server from test/harness/server.ts.
--
-- Usage:
--   local CapiProxy = require("e2e.testharness.proxy")
--   local proxy = CapiProxy.new()
--   local url = proxy:start()
--   -- ... run tests against url ...
--   proxy:stop()

local CapiProxy = {}
CapiProxy.__index = CapiProxy

--- Create a new CapiProxy instance.
-- @return CapiProxy
function CapiProxy.new()
    local self = setmetatable({}, CapiProxy)
    self._process = nil
    self._proxy_url = nil
    self._original_copilot_api_url = nil
    return self
end

--- Launch the proxy server and return its URL.
-- Sets the COPILOT_API_URL environment variable so the CLI uses the proxy.
-- @return string|nil  The proxy URL
-- @return string|nil  Error message
function CapiProxy:start()
    if self._proxy_url then
        return self._proxy_url, nil
    end

    -- Resolve path to the shared harness server relative to this file's location.
    -- From lua/e2e/testharness/ we need to reach test/harness/server.ts
    -- which is at ../../test/harness/server.ts relative to lua/e2e/.
    local sep = package.config:sub(1, 1)
    local is_windows = (sep == "\\")

    -- Build the path to the harness server.
    -- We get the directory of this script to build a reliable path.
    local script_dir = debug.getinfo(1, "S").source:match("^@(.+)[/\\]") or "."
    local server_path = script_dir .. sep .. ".." .. sep .. ".." .. sep .. ".."
        .. sep .. "test" .. sep .. "harness" .. sep .. "server.ts"

    -- Normalize the path for the OS
    if is_windows then
        server_path = server_path:gsub("/", "\\")
    else
        server_path = server_path:gsub("\\", "/")
    end

    -- The cwd for the proxy process should be the test/harness directory
    local harness_dir = script_dir .. sep .. ".." .. sep .. ".." .. sep .. ".."
        .. sep .. "test" .. sep .. "harness"
    if is_windows then
        harness_dir = harness_dir:gsub("/", "\\")
    else
        harness_dir = harness_dir:gsub("\\", "/")
    end

    -- Build the command. On Windows we need shell to find npx.
    local cmd
    if is_windows then
        cmd = string.format('cd /d "%s" && npx tsx "%s"', harness_dir, server_path)
    else
        cmd = string.format('cd "%s" && npx tsx "%s"', harness_dir, server_path)
    end

    -- Spawn the process, reading stdout to capture the listening URL
    local handle = io.popen(cmd, "r")
    if not handle then
        return nil, "Failed to spawn replay proxy process"
    end

    self._process = handle

    -- Read the first line of output to get the listening URL
    local line = handle:read("*l")
    if not line or line == "" then
        handle:close()
        self._process = nil
        return nil, "Failed to read proxy URL from process stdout"
    end

    -- Parse "Listening: http://..." from output
    local url = line:match("Listening:%s*(http://[^%s]+)")
    if not url then
        handle:close()
        self._process = nil
        return nil, "Unexpected proxy output: " .. tostring(line)
    end

    self._proxy_url = url

    -- Save original env and set COPILOT_API_URL so the CLI uses the proxy
    self._original_copilot_api_url = os.getenv("COPILOT_API_URL")
    self:_setenv("COPILOT_API_URL", url)

    return url, nil
end

--- Gracefully shut down the proxy server.
-- Restores the COPILOT_API_URL environment variable.
-- @param skip_writing_cache boolean|nil  If true, proxy won't write captured exchanges
function CapiProxy:stop(skip_writing_cache)
    if not self._process then
        return
    end

    -- Send stop request to the server (best effort)
    if self._proxy_url then
        local stop_url = self._proxy_url .. "/stop"
        if skip_writing_cache then
            stop_url = stop_url .. "?skipWritingCache=true"
        end
        -- Use curl for a simple POST; ignore errors
        local is_windows = (package.config:sub(1, 1) == "\\")
        local curl_cmd
        if is_windows then
            curl_cmd = string.format('curl -s -X POST "%s" >nul 2>&1', stop_url)
        else
            curl_cmd = string.format('curl -s -X POST "%s" >/dev/null 2>&1', stop_url)
        end
        os.execute(curl_cmd)
    end

    -- Close the process handle
    pcall(function() self._process:close() end)
    self._process = nil

    -- Restore the original COPILOT_API_URL
    if self._original_copilot_api_url then
        self:_setenv("COPILOT_API_URL", self._original_copilot_api_url)
    else
        self:_setenv("COPILOT_API_URL", "")
    end

    self._proxy_url = nil
end

--- Configure the proxy with a snapshot file path and working directory.
-- @param file_path string  Path to the snapshot YAML file
-- @param work_dir string   Working directory for the proxy
-- @return boolean success
-- @return string|nil error
function CapiProxy:configure(file_path, work_dir)
    if not self._proxy_url then
        return false, "Proxy not started"
    end

    local is_windows = (package.config:sub(1, 1) == "\\")
    local curl_cmd
    local json_body = string.format('{"filePath":"%s","workDir":"%s"}',
        file_path:gsub("\\", "\\\\"):gsub('"', '\\"'),
        work_dir:gsub("\\", "\\\\"):gsub('"', '\\"'))

    if is_windows then
        curl_cmd = string.format(
            'curl -s -X POST -H "Content-Type: application/json" -d "%s" "%s/config"',
            json_body:gsub('"', '\\"'), self._proxy_url)
    else
        curl_cmd = string.format(
            "curl -s -X POST -H 'Content-Type: application/json' -d '%s' '%s/config'",
            json_body, self._proxy_url)
    end

    local ok = os.execute(curl_cmd)
    if not ok then
        return false, "Failed to configure proxy"
    end
    return true, nil
end

--- Get the proxy URL, or nil if not started.
-- @return string|nil
function CapiProxy:url()
    return self._proxy_url
end

-- ---------------------------------------------------------------------------
-- Internal helpers
-- ---------------------------------------------------------------------------

--- Set an environment variable (cross-platform best effort).
-- Lua doesn't have a built-in setenv; we use os.execute as a workaround
-- for child processes, but for the current process we rely on a C binding
-- if available, or fall back to a no-op with a warning.
-- @param name string
-- @param value string
function CapiProxy:_setenv(name, value)
    -- Try luaposix first
    local ok, posix = pcall(require, "posix.stdlib")
    if ok and posix and posix.setenv then
        posix.setenv(name, value)
        return
    end

    -- Try lua-os-setenv (luarocks: os-setenv)
    local ok2, setenv_mod = pcall(require, "os_setenv")
    if ok2 and setenv_mod then
        setenv_mod.setenv(name, value)
        return
    end

    -- Fallback: on POSIX, export for child processes via shell.
    -- This won't affect the current Lua process's os.getenv(), but the
    -- CLI process will inherit the env from popen.
    -- For testing purposes, we store the value so proxy:url() works.
    if package.config:sub(1, 1) ~= "\\" then
        os.execute(string.format('export %s="%s"', name, value))
    end
end

return CapiProxy
